File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        curr = dummy
        carry = 0
        while l1 or l2 or carry:
            if l1:
                num1 = l1.val
            else:
                num1 = 0
            if l2:
                num2 = l2.val
            else:
                num2 = 0
            temp = num1 + num2 + carry
            resultSum = temp % 10
            carry = temp // 10
            curr.next = ListNode(resultSum)
            if l1:
                l1 = l1.next
            else:
                l1 = None
            if l2:
                l2 = l2.next
            else:
                l2 = None
            curr = curr.next
        return dummy.next
            
            
        

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        int num1 = 0;
        int num2 = 0;
        ListNode* dummy = new ListNode();
        ListNode* temp = dummy;
        while(l1 || l2 || carry){
            if(l1){
                num1 = l1->val;
                l1 = l1->next;
            }
            else{
                num1 = 0;
            }
            if(l2){
                num2 = l2->val;
                l2 = l2->next;
            }
            else{
                num2 = 0;
            }
            int sum = num1 + num2 + carry;
            int result = sum % 10;
            carry = sum / 10;
            ListNode* newNode = new ListNode(result);
            temp->next = newNode;
            temp = temp->next;
        }
        return dummy->next;
    }
};